File: examples/engine/inmemory-go/sql.go
// sql.go — the deliberately tiny mini-SQL parser.
//
// The grammar is intentionally minimal + rigidly canonical so that two
// independent reference implementations (this and inmemory-py) parse identically
// and the golden vectors stay in lockstep. The SAME three regexes are used in both
// languages. Grammar (case-sensitive keywords, single-space separated):
//
//	CREATE TABLE <t> (<col>, <col>, ...)
//	INSERT INTO <t> VALUES (<v>, <v>, ...)
//	SELECT <* | col, col, ...> FROM <t> [WHERE <col> = <val>] [LIMIT <n>]
//
// Identifiers + values are [A-Za-z0-9_]+. Anything else is a parse error →
// INVALID_ARGUMENT at the RPC boundary. SQL fidelity is NOT the point; the engine
// WIRE contract is.
package main

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

type stmt struct {
	kind               string // "create" | "insert" | "select"
	table              string
	cols               []string // create columns / select projection (["*"] = all)
	vals               []string // insert values
	whereCol, whereVal string
	hasWhere           bool
	limit              int64
	hasLimit           bool
}

var (
	reCreate = regexp.MustCompile(`^CREATE TABLE (\w+) \(([\w, ]+)\)$`)
	reInsert = regexp.MustCompile(`^INSERT INTO (\w+) VALUES \(([\w, ]+)\)$`)
	reSelect = regexp.MustCompile(`^SELECT (\*|[\w, ]+) FROM (\w+)( WHERE (\w+) = (\w+))?( LIMIT (\d+))?$`)

	errEmptySQL = errors.New("sql is required")
	errParse    = errors.New("unparseable sql (mini-SQL grammar only)")
)

func errUnknownTable(name string) error { return fmt.Errorf("unknown table %q", name) }

// splitList splits a ", "-separated list (CREATE cols / INSERT vals / projection).
func splitList(s string) []string { return strings.Split(s, ", ") }

func parseSQL(sql string) (stmt, error) {
	sql = strings.TrimSpace(sql)
	if sql == "" {
		return stmt{}, errEmptySQL
	}
	if m := reCreate.FindStringSubmatch(sql); m != nil {
		return stmt{kind: "create", table: m[1], cols: splitList(m[2])}, nil
	}
	if m := reInsert.FindStringSubmatch(sql); m != nil {
		return stmt{kind: "insert", table: m[1], vals: splitList(m[2])}, nil
	}
	if m := reSelect.FindStringSubmatch(sql); m != nil {
		st := stmt{kind: "select", table: m[2]}
		if m[1] == "*" {
			st.cols = []string{"*"}
		} else {
			st.cols = splitList(m[1])
		}
		if m[3] != "" { // " WHERE col = val"
			st.hasWhere, st.whereCol, st.whereVal = true, m[4], m[5]
		}
		if m[6] != "" { // " LIMIT n"
			n, err := strconv.ParseInt(m[7], 10, 64)
			if err != nil {
				return stmt{}, errParse
			}
			st.hasLimit, st.limit = true, n
		}
		return st, nil
	}
	return stmt{}, errParse
}

File: plugins/engine/inmemory-py/sql.py
"""The deliberately tiny mini-SQL parser — Python mirror of the Go reference's
sql.go. The SAME three regexes are used in both languages so the two independent
parsers stay in lockstep on the shared golden vectors.

Grammar (case-sensitive keywords, single-space separated):
    CREATE TABLE <t> (<col>, <col>, ...)
    INSERT INTO <t> VALUES (<v>, <v>, ...)
    SELECT <* | col, col, ...> FROM <t> [WHERE <col> = <val>] [LIMIT <n>]

Identifiers + values are [A-Za-z0-9_]+. Anything else is a parse error.
"""

import re
from dataclasses import dataclass, field
from typing import List, Optional

_CREATE = re.compile(r"^CREATE TABLE (\w+) \(([\w, ]+)\)$")
_INSERT = re.compile(r"^INSERT INTO (\w+) VALUES \(([\w, ]+)\)$")
_SELECT = re.compile(r"^SELECT (\*|[\w, ]+) FROM (\w+)( WHERE (\w+) = (\w+))?( LIMIT (\d+))?$")


class SQLError(Exception):
    """Any mini-SQL problem → INVALID_ARGUMENT at the RPC boundary."""


@dataclass
class Stmt:
    kind: str  # "create" | "insert" | "select"
    table: str
    cols: List[str] = field(default_factory=list)  # create cols / select projection (["*"] = all)
    vals: List[str] = field(default_factory=list)  # insert values
    where_col: Optional[str] = None
    where_val: Optional[str] = None
    has_where: bool = False
    limit: int = 0
    has_limit: bool = False


def _split_list(s: str) -> List[str]:
    return s.split(", ")


def parse_sql(sql: str) -> Stmt:
    sql = sql.strip()
    if not sql:
        raise SQLError("sql is required")
    m = _CREATE.match(sql)
    if m:
        return Stmt(kind="create", table=m.group(1), cols=_split_list(m.group(2)))
    m = _INSERT.match(sql)
    if m:
        return Stmt(kind="insert", table=m.group(1), vals=_split_list(m.group(2)))
    m = _SELECT.match(sql)
    if m:
        st = Stmt(kind="select", table=m.group(2))
        st.cols = ["*"] if m.group(1) == "*" else _split_list(m.group(1))
        if m.group(3):  # " WHERE col = val"
            st.has_where, st.where_col, st.where_val = True, m.group(4), m.group(5)
        if m.group(6):  # " LIMIT n"
            st.has_limit, st.limit = True, int(m.group(7))
        return st
    raise SQLError("unparseable sql (mini-SQL grammar only)")
